### PR TITLE
Handle issues introduced by OTLP gRPC protocol

### PR DESCRIPTION
**Pull request by thpierce**
In this commit, we are handling issues that arise from gRPC. Essentially, if we build gRPC artifacts into our Docker image, it causes the Docker image to only be compatible with applications built using the same Python version. To solve this, we are doing two things: 1) we are removing gRPC artifacts from the docker image and 2) we are changing the default OTLP protocol to be HTTP. If customers attempt to set  the protocol as gRPC for ApplicationSignals, we will set the default endpoint correctly.

Also we are changing Docker image to build with Python 3.11, which is what we were originally doing when we encountered this issue (reference: https://github.com/aws-observability/aws-otel-python-instrumentation/commit/5b3ed74eb8fd93a7810380dcd82234c6028423a4). This is what the upstream does (see [autoinstrumentation/python/Dockerfile](https://github.com/open-telemetry/opentelemetry-operator/blob/b5bb0ae34720d4be2d229dafecb87b61b37699b0/autoinstrumentation/python/Dockerfile)), and having parity here is beneficial to us.

Testing:
* Create `app.py`:
```
from time import sleep
import boto3

try:
    boto3.client('s3').list_buckets()
except Exception:
    sleep(100)

```
* Run `./scripts/build_and_install_distro.sh`
* Run:
```
export OTEL_PYTHON_DISTRO="aws_distro"         
export OTEL_PYTHON_CONFIGURATOR="aws_configurator"
export OTEL_METRICS_EXPORTER="none"
unset OTEL_EXPORTER_OTLP_PROTOCOL
unset OTEL_AWS_APPLICATION_SIGNALS_ENABLED
```
* Run `opentelemetry-instrument python ./app.py`
```
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=4318): Max retries exceeded with url: /v1/traces (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc647cd5e50>: Failed to establish a new connection: [Errno 111] Connection refused'))
```
* Run `export OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf; opentelemetry-instrument python ./app.py`
```
requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=4318): Max retries exceeded with url: /v1/traces (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f20ac96e490>: Failed to establish a new connection: [Errno 111] Connection refused'))
```
* Run `export OTEL_EXPORTER_OTLP_PROTOCOL=grpc; opentelemetry-instrument python ./app.py`
```
Transient error StatusCode.UNAVAILABLE encountered while exporting traces to localhost:4317, retrying in 1s.
```
* Run
```
unset OTEL_EXPORTER_OTLP_PROTOCOL
export OTEL_METRIC_EXPORT_INTERVAL=1000
export OTEL_AWS_APPLICATION_SIGNALS_ENABLED=True
```
* Run `opentelemetry-instrument python ./app.py`
```
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=4316): Max retries exceeded with url: /v1/metrics (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8f6c5aa5b0>: Failed to establish a new connection: [Errno 111] Connection refused'))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=4318): Max retries exceeded with url: /v1/traces (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8f6dd9c9d0>: Failed to establish a new connection: [Errno 111] Connection refused'))
```
* Run `export OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf; opentelemetry-instrument python ./app.py`
```
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=4318): Max retries exceeded with url: /v1/traces (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff5868efdc0>: Failed to establish a new connection: [Errno 111] Connection refused'))
requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=4318): Max retries exceeded with url: /v1/traces (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff5868efdc0>: Failed to establish a new connection: [Errno 111] Connection refused'))
```
* Run `export OTEL_EXPORTER_OTLP_PROTOCOL=grpc; opentelemetry-instrument python ./app.py`
```
Transient error StatusCode.UNAVAILABLE encountered while exporting metrics to localhost:4315, retrying in 1s.
Transient error StatusCode.UNAVAILABLE encountered while exporting traces to localhost:4317, retrying in 1s.
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Dockerfile
Comment (srprash) [inline Dockerfile:20]: The `opentelemetry-exporter-otlp-proto-grpc` installs other dependencies as well. Do you think we should remove those as well? Or just the problematic `grpcio` for now?
IMO, just removing the `grpcio` should be fine but want to double-check.
```
(venv) ➜  aws-otel-python-instrumentation git:(main) ✗ pipdeptree                                                
opentelemetry-exporter-otlp-proto-grpc==1.22.0
├── backoff [required: >=1.10.0,<3.0.0, installed: 2.2.1]
├── Deprecated [required: >=1.2.6, installed: 1.2.14]
│   └── wrapt [required: >=1.10,<2, installed: 1.16.0]
├── googleapis-common-protos [required: ~=1.52, installed: 1.63.0]
│   └── protobuf [required: >=3.19.5,<5.0.0.dev0,!=4.21.5,!=4.21.4,!=4.21.3,!=4.21.2,!=4.21.1,!=3.20.1,!=3.20.0, installed: 4.25.3]
├── grpcio [required: >=1.0.0,<2.0.0, installed: 1.62.2]
├── opentelemetry-api [required: ~=1.15, installed: 1.22.0]
│   ├── Deprecated [required: >=1.2.6, installed: 1.2.14]
│   │   └── wrapt [required: >=1.10,<2, installed: 1.16.0]
│   └── importlib-metadata [required: >=6.0,<7.0, installed: 6.11.0]
│       └── zipp [required: >=0.5, installed: 3.18.1]
├── opentelemetry-exporter-otlp-proto-common [required: ==1.22.0, installed: 1.22.0]
│   ├── backoff [required: >=1.10.0,<3.0.0, installed: 2.2.1]
│   └── opentelemetry-proto [required: ==1.22.0, installed: 1.22.0]
│       └── protobuf [required: >=3.19,<5.0, installed: 4.25.3]
├── opentelemetry-proto [required: ==1.22.0, installed: 1.22.0]
│   └── protobuf [required: >=3.19,<5.0, installed: 4.25.3]
└── opentelemetry-sdk [required: ~=1.22.0, installed: 1.22.0]
    ├── opentelemetry-api [required: ==1.22.0, installed: 1.22.0]
    │   ├── Deprecated [required: >=1.2.6, installed: 1.2.14]
    │   │   └── wrapt [required: >=1.10,<2, installed: 1.16.0]
    │   └── importlib-metadata [required: >=6.0,<7.0, installed: 6.11.0]
    │       └── zipp [required: >=0.5, installed: 3.18.1]
    ├── opentelemetry-semantic-conventions [required: ==0.43b0, installed: 0.43b0]
    └── typing_extensions [required: >=3.7.4, installed: 4.11.0]

```

File: contract-tests/tests/test/amazon/base/contract_test_base.py
Comment (thpierce) [inline contract-tests/tests/test/amazon/base/contract_test_base.py:93]: Note that this works because we install `opentelemetry-exporter-otlp-proto-grpc==1.22.0` in each sample application Docker image: https://github.com/search?q=repo%3Aaws-observability%2Faws-otel-python-instrumentation%20opentelemetry-exporter-otlp-proto-grpc%3D%3D1.22.0&type=code

File: aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_configurator.py
Comment (srprash) [inline aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_configurator.py:282]: the http endpoint for pulse metrics need to be `http://localhost:4316/v1/metrics`.

File: aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_configurator.py
Comment (srprash) [inline aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_configurator.py:308]: Does this mean we can still support grpc protocol if a user specifies it?

File: aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_distro.py
Comment (srprash) [inline aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_distro.py:44]: Curious to why these configs are moved up?

File: contract-tests/tests/test/amazon/base/contract_test_base.py
Comment (srprash) [inline contract-tests/tests/test/amazon/base/contract_test_base.py:93]: Can we provide an explanation why we need to specify `grpc` or contract tests and not leave it to default?

File: Dockerfile
Comment (thpierce) [reply Dockerfile:20]: I think all of those are shared with `opentelemetry-exporter-otlp-proto-http`:
```
> pip show opentelemetry-exporter-otlp-proto-http
Name: opentelemetry-exporter-otlp-proto-http
Version: 1.22.0
Summary: OpenTelemetry Collector Protobuf over HTTP Exporter
Home-page: 
Author: 
Author-email: OpenTelemetry Authors <cncf-opentelemetry-contributors@lists.cncf.io>
License: 
Location: /local/home/thp/.local/lib/python3.9/site-packages
Requires: backoff, deprecated, googleapis-common-protos, opentelemetry-api, opentelemetry-exporter-otlp-proto-common, opentelemetry-proto, opentelemetry-sdk, requests
Required-by: aws-opentelemetry-distro

> pip show opentelemetry-exporter-otlp-proto-grpc
Name: opentelemetry-exporter-otlp-proto-grpc
Version: 1.22.0
Summary: OpenTelemetry Collector Protobuf over gRPC Exporter
Home-page: 
Author: 
Author-email: OpenTelemetry Authors <cncf-opentelemetry-contributors@lists.cncf.io>
License: 
Location: /local/home/thp/.local/lib/python3.9/site-packages
Requires: backoff, deprecated, googleapis-common-protos, grpcio, opentelemetry-api, opentelemetry-exporter-otlp-proto-common, opentelemetry-proto, opentelemetry-sdk
Required-by: aws-opentelemetry-distro
```

No action required.

File: aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_distro.py
Comment (thpierce) [reply aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_distro.py:44]: Ah sorry I must have moved them around and forgot, it's a no-op, safe to ignore. Sorry for the confusion.

File: aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_configurator.py
Comment (thpierce) [reply aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_configurator.py:308]: Yes, assuming `opentelemetry-exporter-otlp-proto-grpc` is installed.

File: contract-tests/tests/test/amazon/base/contract_test_base.py
Comment (thpierce) [reply contract-tests/tests/test/amazon/base/contract_test_base.py:93]: Contract test framework is entirely built around gRPC export (mock collector is a gRPC server), which was the default. Now default is HTTP, so if we leave it as default, tests will all fail.

File: aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_distro.py
Comment (mxiamxia) [inline aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_distro.py:45]: q: do we need to explicitly specify `OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4316` as default endpoint here as well? Will the global SpanExporter also be initiated with the default HTTP endpoint?

File: aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_configurator.py
Comment (srprash) [reply aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_configurator.py:308]: Do we plan to call this out to users in our public docs or somewhere similar?

File: aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_distro.py
Comment (thpierce) [reply aws-opentelemetry-distro/src/amazon/opentelemetry/distro/aws_opentelemetry_distro.py:45]: No, default endpoint is tied to choice of protocol: https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/#otel_exporter_otlp_endpoint & https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-python+4318&type=code

Testing done will show this as well (working on it)

File: contract-tests/tests/test/amazon/base/contract_test_base.py
Comment (srprash) [reply contract-tests/tests/test/amazon/base/contract_test_base.py:93]: I see. Maybe we should put a comment in the code explaining this.

Comment (thpierce) [reply contract-tests/tests/test/amazon/base/contract_test_base.py:93]: There's lots of configuration here without comment. I don't think this needs to block this PR.

**Review by mxiamxia — APPROVED** [review]

**Review by srprash — APPROVED** [review]

**Review by mxiamxia — APPROVED** [review]

**Pull request merged**
